zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreter.java

```diff
@@ -0,0 +1,371 @@
+  private volatile boolean isOpened = false;
+  private volatile boolean isCreated = false;
+
+  /**
+   * Remote interpreter and manage interpreter process
+   */
+  public RemoteInterpreter(Properties properties,
+                           String sessionId,
+                           String className,
+                           String userName) {
+    super(properties);
+    this.sessionId = sessionId;
+    this.className = className;
+    this.userName = userName;
+  }
+
+  public boolean isOpened() {
+    return isOpened;
+  }
+
+  @Override
+  public String getClassName() {
+    return className;
+  }
+
+  public String getSessionId() {
+    return this.sessionId;
+  }
+
+  public synchronized RemoteInterpreterProcess getOrCreateInterpreterProcess() {
+    if (this.interpreterProcess != null) {
+      return this.interpreterProcess;
+    }
+    InterpreterGroup intpGroup = getInterpreterGroup();
+    this.interpreterProcess = intpGroup.getOrCreateInterpreterProcess();
+    synchronized (interpreterProcess) {
+      if (!interpreterProcess.isRunning()) {
+        interpreterProcess.start(userName, false);
+        interpreterProcess.getRemoteInterpreterEventPoller()
+            .setInterpreterProcess(interpreterProcess);
+        interpreterProcess.getRemoteInterpreterEventPoller().setInterpreterGroup(intpGroup);
+        interpreterProcess.getRemoteInterpreterEventPoller().start();
+      }
+    }
+    return interpreterProcess;
+  }
+
+  @Override
+  public void open() {
+    synchronized (this) {
+      if (!isOpened) {
+        // create all the interpreters of the same session first, then Open the internal interpreter
+        // of this RemoteInterpreter.
+        // The why we we create all the interpreter of the session is because some interpreter
+        // depends on other interpreter. e.g. PySparkInterpreter depends on SparkInterpreter.
+        // also see method Interpreter.getInterpreterInTheSameSessionByClassName
+        for (Interpreter interpreter : getInterpreterGroup().getOrCreateSession(
+            userName, sessionId)) {
+          ((RemoteInterpreter) interpreter).internal_create();
+        }
+
+        interpreterProcess.callRemoteFunction(new RemoteInterpreterProcess.RemoteFunction<Void>() {
+          @Override
+          public Void call(Client client) throws Exception {
+            LOGGER.info("Open RemoteInterpreter {}", getClassName());
+            client.open(sessionId, className);
+            // Push angular object loaded from JSON file to remote interpreter
+            synchronized (getInterpreterGroup()) {
+              if (!getInterpreterGroup().isAngularRegistryPushed()) {
+                pushAngularObjectRegistryToRemote(client);
+                getInterpreterGroup().setAngularRegistryPushed(true);
+              }
+            }
+            return null;
+          }
+        });
+        isOpened = true;
+      }
+    }
+  }
+
+  private void internal_create() {
```

Comment:
See the comment here. https://github.com/apache/zeppelin/pull/2422/files#diff-11a09579e097e46c35ebe8ed934a5e4bR110